heat/core/communication.py

```diff
@@ -463,6 +464,161 @@ def Iallgatherv(self, sendbuf, recvbuf, send_axis=0, recv_axis=None):
+            mpi_type = mpi_type.Create_vector(shape[i], 1, strides[i]).Create_resized(0, offsets[i])
+            mpi_type.Commit()
+
+        # Step 2: Receive blocks along the recv axis
+        # Prepare recvcount, senddispls and sendtypes for alltoallw
+        recvcount = np.full((nproc,), obj.shape[0]//nproc)
+        recvcount[:obj.shape[0]%nproc]+=1
+        # size/extent of mpitype = offsets[0]
+        tmp_displs = [0] * nproc
+        tmp_displs[1:] = np.cumsum(recvcount[:-1] )
+        recvdispls = [offsets[0] * d for d in tmp_displs]
+        recvtypes = [mpi_type] * nproc
+
+        return cls.as_mpi_memory(obj), (recvcount, recvdispls), recvtypes
+
+
+    def __alltoall_w(self, sendbuf, recvbuf, send_axis, recv_axis, **kwargs):
+
+        if (recv_axis == send_axis):
+            raise NotImplementedError('AllToAll for same axes not supported. Please choose send_axis and recv_axis to be different.')
+
+        # dummy allocation for *v calls
+        send_counts, send_displs, recv_counts, recv_displs = None, None, None, None,
+
+        # unpack the send buffer
+        if isinstance(sendbuf, tuple):
+            sendbuf, send_counts, send_displs = sendbuf
+        if isinstance(sendbuf, dndarray.DNDarray):
+            sendbuf = sendbuf._DNDarray__array
+        if not isinstance(sendbuf, torch.Tensor) and send_axis != 0:
+            raise TypeError('sendbuf of type {} does not support send_axis != 0'.format(type(sendbuf)))
+
+        # unpack the receive buffer
+        if isinstance(recvbuf, tuple):
+            recvbuf, recv_counts, recv_displs = recvbuf
+        if isinstance(recvbuf, dndarray.DNDarray):
+            recvbuf = recvbuf._DNDarray__array
+        if not isinstance(recvbuf, torch.Tensor) and send_axis != 0:
+            raise TypeError('recvbuf of type {} does not support send_axis != 0'.format(type(recvbuf)))
+
+        # keep a reference to the original buffer object
+        original_recvbuf = recvbuf
+
+        # Send_axis-Permutation: [recv_axis, send_axis, rest ...]
+        axis_permutation = list(range(recvbuf.ndimension()))
+        if(send_axis == 0):
+            if(recv_axis == 1 ):
+                axis_permutation[send_axis], axis_permutation[recv_axis] = recv_axis, send_axis
+            else:
+                axis_permutation[1], axis_permutation[send_axis] = send_axis, 1
+                axis_permutation[recv_axis] = axis_permutation[0]
+                axis_permutation[0] = recv_axis
+
+        else:
+            axis_permutation[0], axis_permutation[recv_axis] = recv_axis, 0
+            axis_permutation[send_axis] = axis_permutation[1]
+            axis_permutation[1] = send_axis
+
+        sendbuf = sendbuf.permute(*axis_permutation)
+        recvbuf = recvbuf.permute(*axis_permutation)
+
+        # prepare buffer objects
+        nproc = self.size
+        #if sendbuf is not MPI.IN_PLACE: --> Can we allow IN_Place opperation? (Buffer config might be wrong
```

Comment:
MPI_IN_PLACE should in principle be allowed

Reply:
yes, but does it WORK in this setting? when the buffer is not specifically prepared?

Reply:
Well you could come up with weird edge cases of having singular items for example. Why one would do this is questionable, but it would also not be problematic to simply allow for MPI_IN_PLACE?

Reply:
Well, no. 
"Some intracommunicator collective operations do not support the ``in place'' option (e.g., MPI_ALLTOALLV)" (source: https://www.mcs.anl.gov/research/projects/mpi/mpi-standard/mpi-report-2.0/node145.htm)
I also came up with an examplary case where IN_PLACE would crash for alltoall:
If you have a [4,5] Tensor that is split among axis 0 across 4 processes, and you do alltoall from axis zero to axis 1, then every rank as as sendbuffer with 5 elements, but for the receivebuffer, rank 0 needs 8 elements, while ranks 1 to 3 need 4 elements.

Reply:
Fine, I see the point, let's drop it